pkg/healthcheck/healthcheck.go

```diff
@@ -414,7 +443,7 @@ func (hc *HealthChecker) addLinkerdVersionChecks() {
 	hc.checkers = append(hc.checkers, &checker{
 		category:    LinkerdVersionCategory,
 		description: "can determine the latest version",
-		fatal:       true,
+		warning:     true,
```

Comment:
This check needs to remain as fatal, unfortunately, since knowing the latest version is a prerequisite for all of the subsequent version checks. If we can't reach versioncheck.linkerd.io to determine the latest version, then we can't validate that any of the components are running the latest version or not.

Also, since the "control plane is up-to-date" check is already marked as a warning, if that check fails because the `/Version` endpoint is unreachable, that won't be reported back to the CLI. Instead, the CLI will ignore the failure, and return the client, even though the client (assumedly) won't be able to connect to the public API.

I recommend reverting this change and adding a separate set of checks, similar to `LinkerdAPIChecks`, that calls the `/Version` endpoint instead of the `/SelfCheck` endpoint. Then we can continue use `LinkerdAPIChecks` for the `linkerd check` and `linkerd dashboard` commands, and we can use the new set of more limited checks (control plane pod exists and `/Version` endpoint returns without error) for all other commands.

Reply:
@klingerf I agree with the reasoning for leaving the version check as a fatal check.

But I think what you're asking for in your last paragraph has already been addressed.
With the last commit we extracted out of `LinkerdAPIChecks` the checks for (using same wording as #1854):
- control plane namespace exists
- can initialize the control plane API client

and added a new check for:
- "linkerd-controller-*" pod is running (not necessarily ready)

Those were put into the new `LinkerdExistenceChecks` (which doesn't include `/SelfCheck`) that are used by all the CLI commands but `linkerd check` and `linkerd dashboard`. 

`linkerd check` and `linkerd dashboard` continue making the same set of strict checks since they use both `LinkerdExistenceChecks` and `LinkerdAPIChecks`.

Another change was that, before, only `linkerd check` was using `LinkerdVersionChecks` (containing `/Version` and the other related checks). Now all the CLI commands use that as well.

@grampelberg might wanna want to chime in since these changes were really his :slightly_smiling_face:

Reply:
@klingerf Yup, that fatal -> warning change was mine. The intent was to allow cli commands to run if versioncheck.linkerd.io fails, but you're correct that we should just not run the check, and for the `linkerd check` and `dashboard` cases it should fatally fail the subsequent version checks.

@alpeb's correct that we've already broken out a limited set of checks for all commands except `dashboard` and `check`, specifically:
- `KubernetesAPIChecks`
- `LinkerdExistenceChecks`
- `LinkerdVersionChecks`

It sounds like we should not perform `LinkerdVersionChecks` for any commands except `dashboard` and `check`. We can do that by modifying `validatedPublicAPIClient` as follows:

```go
checks := []healthcheck.Checks{
	healthcheck.KubernetesAPIChecks,
	healthcheck.LinkerdExistenceChecks,
}

if apiChecks {
	checks = append(checks, healthcheck.LinkerdAPIChecks)
	checks = append(checks, healthcheck.LinkerdVersionChecks)
}
```

...and then introduce a new type of `/Version` check (which I think you had in your initial PR), and include that as part of `LinkerdExistenceChecks`, something like:

```go
func GetServerVersion(apiClient pb.ApiClient) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	rsp, err := apiClient.Version(ctx, &pb.Empty{})
	if err != nil {
		return nil, err
	}

	return  rsp.GetReleaseVersion(), nil
}
```

... and then we could refactor `CheckServerVersion` to use this new `GetServerVersion` function.

For reference, here's the current `linkerd check` output for this branch:
```
kubernetes-api: can initialize the client..................................[ok]
kubernetes-api: can query the Kubernetes API...............................[ok]
kubernetes-api: is running the minimum Kubernetes API version..............[ok]
linkerd-existence: control plane namespace exists..........................[ok]
linkerd-existence: controler pod is running................................[ok]
linkerd-existence: can initialize the client...............................[ok]
linkerd-api: control plane pods are ready..................................[ok]
linkerd-api: can query the control plane API...............................[ok]
linkerd-api[kubernetes]: control plane can talk to Kubernetes..............[ok]
linkerd-api[prometheus]: control plane can talk to Prometheus..............[ok]
linkerd-api: no invalid service profiles...................................[ok]
linkerd-version: can determine the latest version..........................[warning] -- Unsupported version channel: git
linkerd-version: cli is up-to-date.........................................[warning] -- is running version 0c7049f0 but the latest version is
linkerd-version: control plane is up-to-date...............................[warning] -- is running version 18.12.4 but the latest version is
```

Reply:
Ok great, thanks for the additional info. The updated approach that @siggy laid out makes sense to me.

I'm sorry the code is so convoluted. To help with my own understanding, I made a list of all of the checks that we run when building the public API client that we use when executing the `linkerd stat` command (for example). As this branch is currently written, they are as follows:

* kubernetes-api: can initialize the client
* kubernetes-api: can query the Kubernetes API
* linkerd-existence: control plane namespace exists
* linkerd-existence: controller pod is running
* linkerd-existence: can initialize the client
* linkerd-version: can determine the latest version
* linkerd-version: cli is up-to-date
* linkerd-version: control plane is up-to-date

Those last 3 "linkerd-version" checks will never result in an error message, however, since they are all marked as warnings. The [exitOnError func](https://github.com/linkerd/linkerd2/blob/master/cli/cmd/root.go#L106) that we use when running the checks explicitly ignores all failed checks that are marked as warnings, so they're effectively a no-op.

In order to bring those checks in line with what's described in https://github.com/linkerd/linkerd2/issues/1854#issuecomment-447976710, we should change them as follows:

* kubernetes-api: can initialize the client
* kubernetes-api: can query the Kubernetes API
* linkerd-existence: control plane namespace exists
* linkerd-existence: controller pod is running
* linkerd-existence: can initialize the client
* __linkerd-existence: can successfully call the `/Version` endpoint__
* ~~linkerd-version: can determine the latest version~~
* ~~linkerd-version: cli is up-to-date~~
* ~~linkerd-version: control plane is up-to-date~~

And I agree that updating `addLinkerdExistenceChecks` to include that one additional check and removing the call to `addLinkerdVersionChecks` should do the trick.